Fix compile issue with WebKit header module

pvollan opened the pull request:
#### 7932eaf523195b030b13cbecedbeff8117f5ce25
<pre>
Fix compile issue with WebKit header module
<a href="https://bugs.webkit.org/show_bug.cgi?id=263289">https://bugs.webkit.org/show_bug.cgi?id=263289</a>
rdar://107568693

Reviewed by Elliott Williams and Brent Fulgham.

WTF headers should not be included in header modules. Make the XPC endpoint headers project headers.

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Shared/Cocoa/XPCEndpoint.h:
* Source/WebKit/Shared/Cocoa/XPCEndpointClient.h:
* Source/WebKit/Shared/Cocoa/XPCEndpointClient.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Tests/WebKit/XPCEndpoint.mm:

Canonical link: <a href="https://commits.webkit.org/269542@main">https://commits.webkit.org/269542@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59c73d605674881dd1b634f8bac54fce3a477bf8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22841 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/798 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23930 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24750 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21141 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23100 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1891 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23368 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23082 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/372 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19813 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25604 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20694 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20940 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24750 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/379 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/302 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5449 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/684 "Built successfully") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/459 "Failed to checkout and rebase branch from PR 19196") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham reviewed (approved):
Always nice to forward-declare when possible. r=me

ian-twilightcoder commented on Source/WebKit/Shared/Cocoa/XPCEndpointClient.h:
```diff
@@ -30,7 +30,18 @@
 #include <wtf/OSObjectPtr.h>
-#include <wtf/spi/darwin/XPCSPI.h>
+
+#ifndef __XPC_H__
```
This doesn't work in modules, if the client includes xpc.h before WebKit, `__XPC_H__` still won't be defined because modules build with a clean preprocessor environment.

ian-twilightcoder commented on Source/WebKit/Shared/Cocoa/XPCEndpointClient.h:
```diff
@@ -30,7 +30,18 @@
+
+#ifndef __XPC_H__
+#if OS_OBJECT_USE_OBJC
+OS_OBJECT_DECL(xpc_object);
```
This also doesn't work, you can't redeclare API from a different module, that will result in an error if both modules are seen.

pvollan replied:
Ah, yes, that is a good point. We cannot assume the XPC header will not have been included at this point.

Thanks for reviewing!

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -17494,17 +17497,6 @@
-				650994D028C942B7006EE6DB /* XPCServiceMain.mm in Sources */,
-			);
-			runOnlyForDeploymentPostprocessing = 0;
-		};
```
This is removing the sources build phase for the WebContent.Crashy target, which doesn't seem at all related.

pvollan replied:
Good catch, will fix!

emw-apple commented on Source/WebKit/Platform/ExtraPrivateSymbolsForTAPI.h:
```diff
@@ -25,8 +25,28 @@
+
+}
+#endif
+
```
Instead of having to check this in and keep it in sync, it might be better to pass XPCEndpoint{,Client}.h has additional arguments to tapi. You could edit OTHER_TAPI_FLAGS in `WebKit.xcconfig` to pass:

```
-extra-private-header $(SRCROOT)/Shared/Cocoa/XPCEndpoint.h -extra-private-header $(SRCROOT)/Shared/Cocoa/XPCEndpointClient.h
```

pvollan replied:
Ah, good point, will fix!

Thanks for reviewing!

ian-twilightcoder replied:
If you're using Xcode 14.3 or later, it's better to set `TAPI_EXTRA_PRIVATE_HEADERS`. But then it looks like this is actually a project header so really you should be using `TAPI_EXTRA_PROJECT_HEADERS`

pvollan replied:
Ah, good catch, Ian! I can look at that in a follow-up patch.

Thanks for reviewing!

emw-apple commented on Source/WebKit/Shared/Cocoa/XPCEndpoint.h:
```diff
@@ -27,20 +27,21 @@
 public:
-    WK_EXPORT XPCEndpoint();
-    WK_EXPORT virtual ~XPCEndpoint() = default;
+    XPCEndpoint() = default;
```
Why move the construction logic into `initialize()`? If it was to make tapi happy, maybe the above can help.

pvollan replied:
This was just to reduce the number of exported symbols. With the tapi changes above, I think this change can be reverted.

emw-apple commented on Source/WebKit/Shared/Cocoa/XPCEndpoint.h:
```diff
@@ -27,16 +27,17 @@
-#include "WKDeclarationSpecifiers.h"
+#include <WebKit/WKBase.h>
 #include <wtf/OSObjectPtr.h>
+
```
I'm assuming these blank lines are intentional, or something check-webkit-style asked for?

pvollan replied:
Not intentional, fixed in latest patch.

Thanks for reviewing!

emw-apple reviewed (approved):
Nice! This looks great to me.

pvollan committed:
Fix compile issue with WebKit header module
https://bugs.webkit.org/show_bug.cgi?id=263289
rdar://107568693

Reviewed by Elliott Williams and Brent Fulgham.

WTF headers should not be included in header modules. Make the XPC endpoint headers project headers.

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Shared/Cocoa/XPCEndpoint.h:
* Source/WebKit/Shared/Cocoa/XPCEndpointClient.h:
* Source/WebKit/Shared/Cocoa/XPCEndpointClient.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Tests/WebKit/XPCEndpoint.mm:

Canonical link: https://commits.webkit.org/269542@main
